### PR TITLE
Fix model offsets for entities

### PR DESCRIPTION
**Pull request by kokosha**
- Correct the origin and set it to z = 1.0.
- Adjust small offset in entities.
- Change to using floats in the rendering process.
- Remove unnecessary returns and casts.

### CODE REVIEW

File: korangar/src/loaders/animation/mod.rs [hunk 1]
```diff
@@ -224,59 +222,55 @@ impl AnimationLoader {
...
+                min_top = min(min_top, frame.offset.y - center_y);
+                max_bottom = max(max_bottom, frame.offset.y + (frame.size.y - 1 - center_y));
+            });
 
+            // From here, the code start not using pixel coordinates (center of pixel
+            // coordinates), but instead pixel border + pixel center
+            // coordinates. This also means that we start using float
+            // numbers.
+            // |x|x|x| - x is the pixel center and | is the pixel border.
+            // `x` is in have number format of X.0 and `|`` is in the form X.5.
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:240]: This part sounds odd in english.

File: korangar/src/loaders/animation/mod.rs [hunk 2]
```diff
@@ -224,59 +222,55 @@ impl AnimationLoader {
...
 
                 #[cfg(feature = "debug")]
                 get_cross_matrix(frame);
                 for frame_part in frame.frame_parts.iter_mut() {
-                    let frame_top_left = frame.offset - (frame.size - Vector2::new(1, 1)) / 2;
-                    let frame_part_top_left = frame_part.offset - (frame_part.size - Vector2::new(1, 1)) / 2;
+                    let frame_size = frame.size.map(|val| val as f32);
+                    let frame_origin = frame.offset.map(|val| val as f32);
+
+                    // The offset is to move to pixel corner.
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:252]: This also sounds strange.

File: korangar/src/loaders/animation/mod.rs [hunk 3]
```diff
@@ -224,59 +222,55 @@ impl AnimationLoader {
...
+                    let frame_top_left = Vector2::new(-frame.size.x as f32 / 2.0, -frame.size.y as f32 + 0.5);
+
+                    let frame_part_top_left_shift = -((frame_part.size - Vector2::new(1, 1)) / 2).map(|val| val as f32);
+                    let frame_part_offset = frame_part.offset.map(|val| val as f32);
+                    // The offset is to move from the pixel center to pixel corner.
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:257]: This also sounds strange.

File: korangar/src/loaders/animation/mod.rs [hunk 4]
```diff
@@ -224,59 +222,55 @@ impl AnimationLoader {
...
+                        (frame_origin + frame_part_offset + frame_part_top_left_shift) - Vector2::<f32>::from_value(0.5);
 
                     // Generate the key points of the frame rectangle.
                     let texture_frame_center = Vector2::new(0.0, 1.0);
-
                     // Generate the key points of the frame part rectangle.
                     // In the variables, we removed the term frame_part,
                     // but we are still working with the frame part.
-                    let new_vector = frame_part.size;
+                    let new_vector = frame_part.size.map(|val| val as f32);
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:266]:
```suggestion
                    let new_vector = frame_part.size.map(|value| value as f32);

```

File: korangar/src/world/animation/mod.rs [hunk 5]
```diff
@@ -113,10 +111,13 @@ impl AnimationData {
...
 
     pub fn calculate_world_matrix(&self, camera: &dyn Camera, frame: &AnimationFrame, entity_position: Point3<f32>) -> Matrix4<f32> {
-        // The vertex position is calculated from the center of image, so we need
-        // to add half of the height.
-        let center_position = Vector2::new(-frame.offset.x as f32, frame.offset.y as f32 + ((frame.size.y - 1) / 2) as f32);
-        let origin = Point3::from_vec(center_position.extend(0.0)) * SPRITE_SCALE / TILE_SIZE;
+        // Offset the image to below the ground by frame.offset.y.
+        // Add 0.5 to change from center of pixel to the lower border of pixel
+        let origin_y = -frame.offset.y as f32 + 0.5;
+        // TBD: Change the entity z coordinate to 0.0.
```
Comment (hasenbanck) [inline korangar/src/world/animation/mod.rs:117]: Makes this please an TODO.

File: korangar/src/loaders/animation/mod.rs [hunk 6]
```diff
@@ -224,59 +222,55 @@ impl AnimationLoader {
...
+                let center_x = (frame.size.x - 1) / 2;
+                min_left = min(min_left, frame.offset.x - center_x);
+                max_right = max(max_right, frame.offset.x + (frame.size.x - 1 - center_x));
+
+                let center_y = (frame.size.y - 1) / 2;
+                min_top = min(min_top, frame.offset.y - center_y);
+                max_bottom = max(max_bottom, frame.offset.y + (frame.size.y - 1 - center_y));
+            });
 
+            // From here, the code start not using pixel coordinates (center of pixel
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:235]: Can you explain in the comment why we do this coordinate shift?

**Review by hasenbanck — CHANGES_REQUESTED** [review]

File: korangar/src/loaders/animation/mod.rs (hunk 4 above)
Comment (kokosha) [reply korangar/src/loaders/animation/mod.rs]: I created a function to change from Vector2<i32> to Vector2<f32>, I think it is better.

File: korangar/src/loaders/animation/mod.rs (hunk 6 above)
Comment (kokosha) [reply korangar/src/loaders/animation/mod.rs:235]: Yeah, I explained better in the code, but the main reason is that you need the pixel border during rendering process.

File: korangar/src/loaders/animation/mod.rs [hunk 7]
```diff
@@ -312,51 +309,43 @@ impl AnimationLoader {
     }
 }
 
+fn vector2_f32(vector: Vector2<i32>) -> Vector2<f32> {
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:312]:
```suggestion
fn vector2_i32_to_f32(vector: Vector2<i32>) -> Vector2<f32> {

```

File: korangar/src/loaders/animation/mod.rs [hunk 8]
```diff
@@ -402,8 +391,8 @@ fn calculate_translation_matrix(
 /// This function converts to the "normalized" coordinates of a frame part
 /// inside the frame's bounding rectangle, defined by the vertices [-1, 0],
 /// [-1, 2], [1, 0], and [1, 2].
-fn convert_coordinates(coordinates: Vector2<i32>, size: Vector2<i32>) -> Vector2<f32> {
-    assert!(size.x != 0 && size.y != 0);
+fn convert_coordinates(coordinates: Vector2<f32>, size: Vector2<f32>) -> Vector2<f32> {
+    assert!(size.x != 0.0 && size.y != 0.0);
     let x = (coordinates.x as f32 / size.x as f32 - 1.0 / 2.0) * 2.0;
```
Comment (hasenbanck) [inline korangar/src/loaders/animation/mod.rs:396]: Please run clippy. I get many linter errors for your PR. Unecessary returns and also casts.

**Review by hasenbanck — CHANGES_REQUESTED** [review]

**Comment by hasenbanck** [comment]
Please squash and rebase as soon as you think you are finished.

**Review by hasenbanck — APPROVED** [review]

**Pull request merged**